File: 283. Move Zeroes.py
# 解法1：使用新的array
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # 创建一个新array
        newArray = [-1] * len(nums)
        # 遍历nums，如果元素非0，放进新array，同时统计0的总数
        totalZero = 0;
        for i in range(len(nums)):
            if nums[i] == 0:
                totalZero += 1
            else:
                newArray[i-totalZero] = nums[i]
        # 把新array的尾部填充满0
        for j in range(len(nums)-totalZero, len(nums)):
            newArray[j] = 0
        # 把新array的值给nums
        for i in range(len(nums)):
            nums[i] = newArray[i]


# 解法2：非0元素往前放，尾部用0填充
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        last_zero = 0
        for i in range(len(nums)):
            if nums[i] != 0:
                nums[last_zero] = nums[i]
                last_zero += 1
        for i in range(last_zero, len(nums)):
            nums[i] = 0


# 解法3：
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        zero = 0  # records the position of "0"
        for i in range(len(nums)):
            if nums[i] != 0:
                nums[i], nums[zero] = nums[zero], nums[i]
                zero += 1

"""
分析：
两个指针，一个遍历元素，另一个先按顺序遍历直到找到第一个0
在这个0后面的元素会出现两种情况：不为0和为0
如果后面的元素不为0，则将二者的值交换，并把指针加1 （实现的是把后面第一个不为0的元素放在第一个0前面）
如果后面的元素为0，则什么都不做（即两个0并排放在一起），继续寻找下一个元素
用一个具体例子来形容：在找到第一个0之前所有元素不动，在找到0之后，遇到非0元素就抽出来放在0前面，遇到0就叠加在一起继续往后推
"""

File: 283. Move Zeroes.java
// 解法1：使用一个新array，对它进行操作，最后用它覆盖nums
class Solution {
    public void moveZeroes(int[] nums) {
        // 创建一个新array
        int[] newArray = new int[nums.length];
        // 遍历nums，如果元素非0，放进新array，同时统计0的总数
        int totalZero = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                totalZero += 1;
            } else {
                newArray[i-totalZero] = nums[i];
            }
        }
        // 把新array的尾部填充满0
        for (int j = nums.length-totalZero; j < nums.length; j ++) {
            newArray[j] = 0;
        }
        // 把新array的值给nums
        for (int i = 0; i < nums.length; i++) {
            nums[i] = newArray[i];
        }
        /*
        以下是错误方法，因为指针虽然变了但本体并未改变：

        // 让nums的指针指向新array
        nums = newArray;

        */
    }
}


// 解法2：将非0元素放置在前，然后尾部用0填充满
class Solution {
    public void moveZeroes(int[] nums) {
        int lastNonZeroFoundAt = 0;
        // If the current element is not 0, then we need to
        // append it just in front of last non 0 element we found.
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                // a = b++: 先 a = b；再 b += 1
                // a = ++b: 先 b += 1；再 a = b
                nums[lastNonZeroFoundAt++] = nums[i];
            }
        }
        // After we have finished processing new elements,
        // all the non-zero elements are already at beginning of array.
        // We just need to fill remaining array with 0's.
        for (int i = lastNonZeroFoundAt; i < nums.length; i++) {
            nums[i] = 0;
        }
    }
}


// 解法3：详见python版解析
class Solution {
    public void moveZeroes(int[] nums) {
        int zero = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                int temp = nums[i];
                nums[i] = nums[zero];
                nums[zero] = temp;
                zero += 1;
            }
        }
    }
}
